File: algo/3.py
# 62ms 책의 풀이에 기반

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        start, max_length = 0, 0
        seen = {}

        for i, char in enumerate(s):
            if char in seen and start <= seen[char]:
                start = seen[char] + 1
            max_length = max(max_length, i - start + 1)
            seen[char] = i
        return max_length


a = Solution()
s = "abcabcbb"
print(a.lengthOfLongestSubstring(s))
s = "bbbbb"
print(a.lengthOfLongestSubstring(s))
s = "pwwkew"
print(a.lengthOfLongestSubstring(s))
s = "aab"
print(a.lengthOfLongestSubstring(s))

File: algo/3.kt
import kotlin.math.max

class Solution {
    fun lengthOfLongestSubstring(s: String): Int {
        var left = 0
        val seen = mutableMapOf<Char, Int>()
        var maxLength = Int.MIN_VALUE
        for ((right, ch) in s.withIndex()) {
            if (seen.contains(ch) && seen.getValue(ch) >= left) {
                left = seen.getValue(ch) + 1
            }
            maxLength = max(maxLength, right - left + 1)
            seen[ch] = right
        }
        return maxLength
    }
}

//val s = "abcabcbb"
//val s = "bbbbb"
val s = "pwwkew"
println(Solution().lengthOfLongestSubstring(s))
